### ISSUE TITLE
const variable placement issue in distributed tensorflow

### ISSUE THREAD

**Issue by YongCHN** [tensorflow/tensorflow#12143]
### Describe the problem
In my model, I use the FTRL Optimizer like below:

`self.optimizer = tf.train.FtrlOptimizer(0.005,
                learning_rate_power=-0.5,
                initial_accumulator_value=0.1,
                l1_regularization_strength=1.0,
                l2_regularization_strength=0.00001)
`

Inside the FtrlOptimizer it will create several const variables for the parameters, such as learning rate, learning rate power, etc. 

When I run the distributed tensorflow job, from the timeline I can see that for each session run I can see that the worker will send the above const variables to all the ps nodes. This is a cost since the variables are const and not needed to sent to ps nodes repeatedly.  

I was wondering is there a way to pin those const variables to the ps and save the transferring cost during each session run.

Thanks.

**Comment by tatatodd** [comment]
@suharshs Do you have any suggestions here?

**Comment by mrry** [comment]
(I think @prb12 might be interested in a solution to this as well....)

**Comment by prb12** [comment]
By 'const variable' (!!), do you mean that the `FtrlOptimizer` creates a `tf.Variable` for learning rate, that is placed on the PS, and is read each step even though the value of that variable never changes?  (I could imagine this would be useful if you ever wanted an adjustable learning rate schedule)

Or... does this produce a `Const` op that is placed on the PS and used on each worker, and the graph partitioner doesn't create a duplicate const on `/cpu:0` of the worker?

**Comment by YongCHN** [comment]
FtrlOptimizer create several const variable such as learning rate, learning rate power, etc, placing at each worker side.

You can see these information when you enabled the variable placement configuration.
```
with sv.managed_session(server.target, config = tf.ConfigProto(log_device_placement=True)) as sess:
```

Then you will see below information.
```
Ftrl/learning_rate_power: (Const): /job:worker/replica:0/task:0/cpu:0
Ftrl/l2_regularization_strength: (Const): /job:worker/replica:0/task:0/cpu:0
Ftrl/l1_regularization_strength: (Const): /job:worker/replica:0/task:0/cpu:0
Ftrl/learning_rate: (Const): /job:worker/replica:0/task:0/cpu:0
```

Then during each session run, I can see that the request from the each ps asking the read of the above variables. I added several logs in the code.
```
3317063:2017-08-10 09:01:01.577890: I tensorflow/core/distributed_runtime/rpc/grpc_worker_service.cc:241] RecvTensorHandlerRaw step_id:127424566250596053  and key:/job:worker/replica:0/task:0/cpu:0;3df08c8c86dccc7c;/job:ps/replica:0/task:3/cpu:0;edge_2045_Ftrl/learning_rate;0:0
```

@prb12 I wish I can place these const variables in the ps nodes, but i don't know how. and in the meantime each ps will need these const variables to apply the gradients.

**Comment by yaroslavvb** [comment]
@YongCHN as a side-comment -- it sounds like you are dealing with Const nodes (there's also Variable node which would be different). I recall seeing Const nodes being cached on devices (ie, fetching same Const by GPU several times would only transfer once), it seems like an omission if they aren't cached by distributed workers.

**Comment by YongCHN** [comment]
@yaroslavvb yes, i am dealing with the Const nodes and I need them be cached in the distributed ps nodes, instead of worker side since these const variables are actually used in the ps nodes.

**Comment by YongCHN** [comment]
Anyone help take a look?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tatatodd** [comment]
@YongCHN Is this still an issue for you?  Have you tried on the latest TensorFlow release?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?
